### PR TITLE
Add "Conversion Checking" flag (ocaml only api)

### PR DESCRIPTION
**Pull request by SkySkimmer**
This was useful when experimenting with the lean importer.

### CODE REVIEW

**Comment by herbelin** [comment]
Why not from my point of view, looks convenient.

**Comment by maximedenes** [comment]
This seems significant enough that it should be discussed at least during a Coq call.

**Comment by herbelin** [comment]
Yes, worth to be discussed. @maximedenes: is your fear that it might be used for "bad" purposes, rather than for "efficient debugging"? Or that it leads to inadvertently "believe" a theory is checked while it is not? Or are you suggesting that we think more about the exact set of deactivated checks?

**Comment by maximedenes** [comment]
Oh, I don't have any specific fear, I just think it is worth mentioning to make sure everybody is ok with it.

**Comment by gares** [comment]
I frankly don't get the implementation, but this is also because the use case is unclear.

It seems it disables a random set of checks. Other checks are still there, and it seems only ML code can actually do something relevant with these checks off, while a final user is likely to hit a type error from pretyping (but not always, since sometimes pretyping delegates to the kernel). Indeed the documentation is too vague, because you cannot possibly explain what this patch does.

Before discussion I think this needs a precise use case.

**Review by JasonGross — DISMISSED** [review]
This should get a test-suite case.  Ideally I'd also like to output in `Print Assumptions` of any code depending on this flag.

**Comment by SkySkimmer** [comment]
I reset to a more sensible version of the code which is what I used for experimenting with the lean importer. The flag is now called "conversion checking" and only allows skipping conversion checks, not random other things.

**Comment by SkySkimmer** [comment]
>This should get a test-suite case. Ideally I'd also like to output in Print Assumptions of any code depending on this flag.

There is both.

**Review by Zimmi48 — COMMENTED** [review]
"checking of conversion problems" is a bit vague for the doc. An example wouldn't hurt.

**Comment by gares** [comment]
It would also help a description of what Qed ensure in this mode (eg a few examples of rejected terms).
I have an hard time myself imagining that myself. Does Qed reject anything at all? Does it accept `Prop Prop` and `fun x : (fun y : Prop => Prop) => Prop`?

**Comment by SkySkimmer** [comment]
Updated doc a bit.

>It would also help a description of what Qed ensure in this mode (eg a few examples of rejected terms).

- universe constraints on polymorphic instances are implied by environment universe constraints (unless type-in-type)
- declared universe constraints are consistent (even with -type-in-type, which you use by declaring no constraints)
- length of polymorphic instances, all universes are declared
- all variables are bound / used globals exist (for non-plugins this means generating/checking the used section variables)
- positivity / termination of (co)fixpoints (unless the relevant flags are toggled)
- some minor stuff like not using primitive ints before the primitive int type is declared

**Comment by gares** [comment]
>     * positivity / termination of (co)fixpoints

I think you should be more specific here, since my understanding is that you type check \Deta \Delta, so you can't tell if fixpoint terminates (since it may call a non terminating term) and also you loose termination of type checking itself, since the current code that checks for termination whf the term.

I'm wondering if it would make more sense to say/implement that the flag disables typechecking altogether. Invariants about universes arity are an implementation detail (if we had a proper quantifier they would amount to type checking as well). Declared constraints... well... I can't possibly see the user case.

Another piece of info a user may want to know (and me as well) is if calling vm_compute (on non checked terms) can crash Coq.

File: doc/sphinx/proof-engine/vernacular-commands.rst
Comment (jfehrle) [inline doc/sphinx/proof-engine/vernacular-commands.rst]: The flags should use the proper names and be hyperlinked to "Guard Checking", "Positivity Checking", "Universe Checking", etc.

Perhaps you would also make the printed output use the official flag names.

image

I would also wonder whether "Print Typing Flags" is really needed.  A user can just do "Test Positivity Checking", etc. for each flag.

File: doc/sphinx/proof-engine/vernacular-commands.rst
Comment (jfehrle) [inline doc/sphinx/proof-engine/vernacular-commands.rst]:
```suggestion
   around them (for instance using :tacn:`exact_no_check`) to use
```

File: doc/sphinx/proof-engine/vernacular-commands.rst
Comment (jfehrle) [inline doc/sphinx/proof-engine/vernacular-commands.rst]:
```suggestion
   this flag.
```

**Review by jfehrle — COMMENTED** [review]
Wording suggestions and one substantive comment--all for the documentation.

**Comment by SkySkimmer** [comment]
What is this PR waiting for?

**Review by ppedrot — COMMENTED** [review]
This looks much more dangerous to me than the set of unsafe flags we currently have available in the Coq implementation. So far, the only way you could get in trouble would be a non-terminating term. With this patch, you can magically cast stuff in the style of `Obj.magic`. In particular, you can seriously mess with the invariants of the VM, and probably trigger a lot of horrendous behaviours like arbitrary code execution. I think that this PR should be more thoroughly discussed before being merged.

**Comment by maximedenes** [comment]
Oh yes, @ppedrot is right. It will be easy to make Coq segfault with the flag on, right @SkySkimmer ?

If so, it seems like a dealbreaker to me.

**Comment by SkySkimmer** [comment]
You can already mess with the VM using exact_no_check though.

**Comment by herbelin** [comment]
@skyskimmer: is there a way to evaluate how this flag would be more segfault-risky than other no-check features? Or maybe, a solution would be to have the flag modifiable only in the ML code, forcing a deliberate recompilation in full awareness to deactivate the conversion check?

**Comment by gares** [comment]
I still fail to to see the use case for a regular user, especially if it may crash Coq. Maybe this should be available only via loading ML code (assuming the ML code is careful to only inject "safe" term into the env)

**Comment by Zimmi48** [comment]
Dear PR author(s) and contributors, it seems to us this PR may need additional work beyond the 8.12 branching deadline on May 15th.

We are thus removing the 8.12+beta1 milestone; please retarget as appropriate, and also consider updating labels to reflect current status, moving the PR to draft status, etc.

We apologize in advance if we misunderstood the PR status.

**Comment by Zimmi48** [comment]
It still OK to add this back into the milestone if it can be finished in time (@coq/kernel-maintainers).

**Comment by ppedrot** [comment]
I don't think it's a good idea to rush this PR. It's clearly touching a very sensitive area. Unfortunately there was not a lot of activity around it, hence the delay...

**Comment by SkySkimmer** [comment]
>Maybe this should be available only via loading ML code

ie removing the goption declaration? I can do that. Would other @coq/kernel-maintainers agree with working that way?

**Comment by ppedrot** [comment]
@SkySkimmer this PR is also affected by #11477, isn't it?

**Comment by SkySkimmer** [comment]
>@SkySkimmer this PR is also affected by #11477, isn't it?

Currently I have the checker ignore the flag so no need to report it IMO.

**Review by maximedenes — COMMENTED** [review]
How about discussing this in today's Coq call? @SkySkimmer, would you mind adding it to the topics?

**Comment by SkySkimmer** [comment]
Coq call: instead of this let's have an API to add trusted definitions directly.

**Pull request closed without merging**
